### PR TITLE
Unit name generator using context-free grammar

### PR DESCRIPTION
**Pull request by Dugy**
Added a new way to generate unit names, using rules of a context-free grammar. It is much more configurable than the current system, so it's possible to prevent extremely long or otherwise funny and awkward names. If it can't find its entries in the race definition, it falls back to the old Markov chains.

### CODE REVIEW

File: src/utils/context_free_grammar_generator.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+ *  @file
+ *  Algorithm to generate names using a context-free grammar, which allows more control
+ *  than the usual Markov chain generator
+ */
+
+#include "context_free_grammar_generator.hpp"
+#include "../log.hpp"
+
+context_free_grammar_generator::context_free_grammar_generator() :
+initialized_(false)
```
Comment (CelticMinstrel) [inline src/utils/context_free_grammar_generator.cpp]: I think this should be indented.

**Comment by CelticMinstrel** [comment]
Looks pretty good! Your indentation seems to be all wrong though - it's using spaces, where everything else is tabs.

Also, I'm not sure I like the syntax. Since it can be described as simple key-value, why create a whole new parser for the grammar? It could be represented as normal WML instead. I'll take the drake names as an example - you currently have this:

```wml
    male_name_generator= _ <<
main={prefix}{suffix}|{prefix}’{suffix}|{prefix}{suffix} {short_name}
short_name=Ohm|Ihn|Ohn|Ka|Murr|Kon|Elch|Ko|Græt|Egar
prefix=Ga|Ge|Gel|Gi|Go|Gra|Gri|Ka|Ko|Kla|Mar|Di|Mer|Mor|Oma|Re|Vel|Vash|Ver|Vla
suffix=ash|kré|lsh|lin|rutin|shinar|rushi|ka|ash|ag|all|ash|slar|li|dian|ritos|rusté|shitt|nick|rath|ditt|kon|mad|tick|nir
>>
    female_name_generator= _ <<
main={prefix}{suffix}|{prefix}’{suffix}|{prefix}{suffix} {short_name}
prefix=Al|Ali|Ash|Aug|Auug|Da|De|Dem|Din|Ol|Oma|Orri|Oshi|Val|Ver|Vel|Vush
suffix=en|nash|irt|glann|ll|ge|lin|nitt|tin|tar|lisk|rida|grra|sill|la|mell
short_name=Mal|Ka|Emth|Dar|Eshi
>>

But I think it would be cleaner to do it like this:

```wml
    [male_name_generator]
        main= _ "[prefix][suffix]|[prefix]’[suffix]|[prefix][suffix] [short_name]"
        short_name= _ "Ohm|Ihn|Ohn|Ka|Murr|Kon|Elch|Ko|Græt|Egar"
        prefix= _ "Ga|Ge|Gel|Gi|Go|Gra|Gri|Ka|Ko|Kla|Mar|Di|Mer|Mor|Oma|Re|Vel|Vash|Ver|Vla"
        suffix= _ "ash|kré|lsh|lin|rutin|shinar|rushi|ka|ash|ag|all|ash|slar|li|dian|ritos|rusté|shitt|nick|rath|ditt|kon|mad|tick|nir"
    [/male_name_generator]
    [female_name_generator]
        main= _ "[prefix][suffix]|[prefix]’[suffix]|[prefix][suffix] [short_name]"
        prefix= _ "Al|Ali|Ash|Aug|Auug|Da|De|Dem|Din|Ol|Oma|Orri|Oshi|Val|Ver|Vel|Vush"
        suffix= _ "en|nash|irt|glann|ll|ge|lin|nitt|tin|tar|lisk|rida|grra|sill|la|mell"
        short_name= _ "Mal|Ka|Emth|Dar|Eshi"
    [female_name_generator]
```

Admittedly, that would make it less flexible for the translators, preventing them from introducing their own nonterminals. I'm not sure whether or not that limitation would be worth the cleaner syntax. I think I'd need someone who speaks a non-Western language to chime in.

Aside from that, I think it would be better to use a different character than curly braces to avoid the need for the <<angled strings>> which many people probably associate with Lua. (I'll admit that there's a somewhat selfish motive for this too - my WML syntax highlighter currently assumes any such string contains Lua code.)

**Comment by gfgtdf** [comment]
this code seems to use rand() for name generation. This means that the name will be different on different cleints in mp even if the have the same language. This doesn't cause oos errors, but afaik mattsc explicitly fixed this in the previous code (https://github.com/wesnoth/wesnoth/commit/071b7550f9cf1fafb0d46a7c9f3704de0d8bc737), so it mostlikeley was important to him.

**Comment by CelticMinstrel** [comment]
Another thing - is there any reason to have both Markov lists and a grammar for a given race? If not, you could introduce a new key that selects which type of generator to use, and then just reuse male_names and female_names for the CFG generator.

**Comment by Dugy** [comment]
1. (indentation) Sorry, I have forgotten that the C++ rules of indentation are better than the WML ones.
2. (custom syntax) It's using entirely its own syntax because mixing it with WML would limit the possibilities of translation (trying to prevent some awkward letter groups from appearing in some language without limiting the number of possibilities would require additional derivation rules) and make it slightly harder to test with external tools. The <<angled string>> is not marked as translatable (while this must be translatable) when used in lua, so that would not do much harm to your WML syntax highlighter.
3. (rand() usage) I have no idea why would anyone need that, but I can add it.
4. (remaining Markov list) They are not needed, but I did not want to remove them as I think they have some historical value.
5. Any idea why does travis clang fail? None of the check fail reports seems related to my changes in any way.

**Comment by Dugy** [comment]
1. (indentation) Sorry, I have forgotten that the C++ rules of indentation are better than the WML ones.
2. (custom syntax) It's using entirely its own syntax because mixing it with WML would limit the possibilities of translation (trying to prevent some awkward letter groups from appearing in some language without limiting the number of possibilities would require additional derivation rules) and make it slightly harder to test with external tools. The <> is not marked as translatable (while this must be translatable) when used in lua, so that would not do much harm to your WML syntax highlighter.
3. (rand() usage) I have no idea why would anyone need that, but I can add it.
4. (remaining Markov list) They are not needed, but I did not want to remove them as I think they have some historical value.
5. Any idea why does travis clang fail? None of the check fail reports seems related to my changes in any way.

**Comment by AI0867** [comment]
Have you checked the number of names that can be generated with the new set of expressions versus the old list? I haven't run the numbers, but I suspect the number is much lower, and with the number of units recruited during the average campaign, the birthday problem is going to pop up.

**Comment by Dugy** [comment]
Dwarvish, Wose and Troll names are always composed of at least three parts, so there will be thousands of possibilities.
Human, Elvish, Merfolk, Khalifate, Drake and Lizard names can be shorter, which means that there is a good chance that a two-part variant is picked, where the number of possibilities is only several hundreds.
Orcish and Naga are always made of two parts, which means there are only several hundreds of possibilities.
Gryphon names have only a few dozens of possibilities, but they were quite limited in names anyway.

I don't know how many names are needed, I can't estimate how many variants the original code can make. I might change the rules to increase the variety wherever necessary. It's doable. I suppose it might be necessary for orcs, because SotBE needs recruiting a lot of them.

**Comment by AI0867** [comment]
The birthday paradox says that the probability of collisions goes as the square root of the number, meaning that thousands and hundreds goes to 30s and 10s of expected generated names before a collision.

While the new syntax is clearly much more powerful, and allows one to define much better rules regarding name generation, the old rules create a much larger variety, with a lot less work.

So while I think the context-free syntax could be quite useful, I don't think the name generation for the core races should be changed like this without further analysis. (though it probably wouldn't hurt the gryphons too much)


As for the tests, in one run there is now a core dump in the test executable. I have no idea what's going on there.

**Comment by CelticMinstrel** [comment]
>2. (custom syntax) It's using entirely its own syntax because mixing it with WML would limit the possibilities of translation (trying to prevent some awkward letter groups from appearing in some language without limiting the number of possibilities would require additional derivation rules) and make it slightly harder to test with external tools. The <> is not marked as translatable (while this must be translatable) when used in lua, so that would not do much harm to your WML syntax highlighter.

Hmm, okay. I still think it would be nicer to not use {} in the syntax, but you do have a point that I could use the translation mark to distinguish Lua from CFG grammars.

>3. (rand() usage) I have no idea why would anyone need that, but I can add it. EDIT: The outcome of the commit is not visible in my code, I can't learn if, when or why was it removed because the file was moved into a different folder since then.

Huh?

> 4. (remaining Markov list) They are not needed, but I did not want to remove them as I think they have some historical value. The function generate_name() falls back to it only if it can't find the context-free grammar code.

I don't think there's any need to leave things in the code for historical value when everything is still there in the repository history. Of course we can't remove the Markov generator itself, since it's used by addons and the map generator (though you could change the latter, if you want), but I don't think it's necessary to keep Markov lists in the data that are not even being used. I guess the only real advantage of leaving them is to have an addon compatible with both 1.12 and the new CFG grammar, but that can be done with preprocessor conditions too.

In case I wasn't clear enough, I'm now suggesting this syntax (again using the drakes as an example):

```wml
    name_generator="context_free" # Default value is "markov"
    male_names= _ <<
main={prefix}{suffix}|{prefix}’{suffix}|{prefix}{suffix} {short_name}
short_name=Ohm|Ihn|Ohn|Ka|Murr|Kon|Elch|Ko|Græt|Egar
prefix=Ga|Ge|Gel|Gi|Go|Gra|Gri|Ka|Ko|Kla|Mar|Di|Mer|Mor|Oma|Re|Vel|Vash|Ver|Vla
suffix=ash|kré|lsh|lin|rutin|shinar|rushi|ka|ash|ag|all|ash|slar|li|dian|ritos|rusté|shitt|nick|rath|ditt|kon|mad|tick|nir
>>
    female_names= _ <<
main={prefix}{suffix}|{prefix}’{suffix}|{prefix}{suffix} {short_name}
prefix=Al|Ali|Ash|Aug|Auug|Da|De|Dem|Din|Ol|Oma|Orri|Oshi|Val|Ver|Vel|Vush
suffix=en|nash|irt|glann|ll|ge|lin|nitt|tin|tar|lisk|rida|grra|sill|la|mell
short_name=Mal|Ka|Emth|Dar|Eshi
>>
```

(Possibly also with [] instead of {}, or something.)

> 5. Any idea why does travis clang fail? None of the check fail reports seems related to my changes in any way.

I have no idea, and I wouldn't worry about it too much. It won't prevent this from being accepted.

**Comment by Dugy** [comment]
@AI0867 I might use some random wovel/consonnant group nonterminals to randomise letters like d/t/th, a/o/e, s/sh/ss and such, that should massively increase the number of possible combinations.

@CelticMinstrel I might replace the { sign with [, but I always considered the << quotation >> to mark that some non-wml source code is inserted. If it is just me, then it's okay, I can change it.
Your suggestion for the change for the WML syntax does not look like a good idea to me, if you have an add-on used both on versions before this commit and after it (which may be the case for a year or two), you will have to bother with version check preprocessor invocations. Also, they key name, male_names, looks like a list of names while the grammar is something significantly different from a list of names.

**Comment by CelticMinstrel** [comment]
Well, formulas (like FormulaAI or formula= in SUF) are usually enclosed in quotes, not angled brackets, so not all non-WML source code appears in angled brackets. I guess it's okay if it stays this way; using a different char is just my personal preference.

I don't personally see a problem with requiring preprocessor checks for addons used in both versions, but your argument about the implication of the "male_names" key does make sense... though it kinda applie to the Markov generator too to a lesser extent (it's a list of names, but isn't used like a simple list).

**Comment by Dugy** [comment]
@CelticMinstrel I have chosen these brackets because their meaning is in some way similar to macro substitution, they also mark some section that is to be substituted. Square brackets are usually used for array indexes, which is quite different.

**Comment by ln-zookeeper** [comment]
It's hard to judge how the output of each new generator for each race would compare to the old ones, so it'd be very convenient to see a pile of names from both side-by-side. Considering how many there are, it might be more convenient to do that in another PR, separate from the technical issues.

**Comment by Dugy** [comment]
@ln-zookeeper The new system's probabilities are easy to calculate or list, for the old one it's worse, as you said, I think that it would be easier to make as small edit that would make it dump 100 000 names when initialising (totally trivial edit) and make some statistics over these files? The change itself is much easier than the whole pull request thing. Adding one line and recompiling is faster than the whole git thing.

**Comment by Dugy** [comment]
Measurement of the current name generator's probabilities:
https://forums.wesnoth.org/viewtopic.php?f=12&t=43792&p=595791#p595791

**Comment by Dugy** [comment]
Okay, indentation fixed, increased the randomness of names for the more common races so that no race has a higher chance (easily estimated) that namesakes are found than with the markov generator (got from experiment). The syntax disagreements were settled. I can commit it now, right?

**Comment by CelticMinstrel** [comment]
You can push (possibly force-push) your updates to your repo whenever you want. We do need to be able to see it, after all.

**Comment by gfgtdf** [comment]
you should merge the fix intention commit into the previous commit.

**Pull request merged**